tests/SelfTest/UsageTests/MatchersRanges.tests.cpp

```diff
@@ -548,3 +554,210 @@ TEST_CASE( "The quantifier range matchers support types with different types ret
+            CHECK_THAT( array_int_a, ElementsAre( vector_int_a ) );
+            CHECK_THAT( array_int_a, UnorderedElementsAre( vector_int_a ) );
+        }
+        SECTION( "Two equal containers of different container types "
+                    "(differ in array N)" ) {
+            const std::array<int, 3> array_int_3{ { 1, 2, 3 } };
+            const std::array<int, 4> array_int_4{ { 1, 2, 3, 4 } };
+            CHECK_THAT( array_int_3, !ElementsAre( array_int_4 ) );
+            CHECK_THAT( array_int_3, !UnorderedElementsAre( array_int_4 ) );
+        }
+        SECTION( "Two equal containers of different container types and value "
+                    "types" ) {
+            const std::array<int, 3> array_int_a{ { 1, 2, 3 } };
+            const std::vector<int> vector_char_a{ 1, 2, 3 };
+            CHECK_THAT( array_int_a, ElementsAre( vector_char_a ) );
+            CHECK_THAT( array_int_a, UnorderedElementsAre( vector_char_a ) );
+        }
+        SECTION( "Two equal containers, one random access, one not" ) {
+            const std::array<int, 3> array_int_a{ { 1, 2, 3 } };
+            const std::list<int> list_char_a{ 1, 2, 3 };
+
+#ifdef CONCEPT_TESTING
+            // Verify these types really are different in random access nature
+            static_assert( std::random_access_iterator<decltype(array_int_a)::iterator> );
+            static_assert( !std::random_access_iterator<decltype(list_char_a)::iterator> );
+#endif
+
+            CHECK_THAT( array_int_a, ElementsAre( list_char_a ) );
+            CHECK_THAT( array_int_a, UnorderedElementsAre( list_char_a ) );
+        }
+    }
+
+    SECTION( "Value type" ) {
+        SECTION( "Two equal containers of different value types" ) {
+            const std::vector<int> vector_int_a{ 1, 2, 3 };
+            const std::vector<char> vector_char_a{ 1, 2, 3 };
+            CHECK_THAT( vector_int_a, ElementsAre( vector_char_a ) );
+            CHECK_THAT( vector_int_a, UnorderedElementsAre( vector_char_a ) );
+        }
+        SECTION( "Two non-equal containers of different value types" ) {
+            const std::vector<int> vector_int_a{ 1, 2, 3 };
+            const std::vector<char> vector_char_b{ 1, 2, 2 };
+            CHECK_THAT( vector_int_a, !ElementsAre( vector_char_b ) );
+            CHECK_THAT( vector_int_a, !UnorderedElementsAre( vector_char_b ) );
+        }
+    } 
+
+
+    SECTION( "Custom predicate" ) {
+
+        auto close_enough = []( int lhs, int rhs ) {
+            return std::abs( lhs - rhs ) <= 1;
+        };
+
+        SECTION( "Two equal non-empty containers (close enough)" ) {
+            const std::vector<int> vector_a{ { 1, 2, 3 } };
+            const std::array<char, 3> array_a_plus_1{ { 2, 3, 4 } };
+            CHECK_THAT( vector_a,
+                        ElementsAre( array_a_plus_1, close_enough ) );
+            CHECK_THAT( vector_a,
+                        UnorderedElementsAre( array_a_plus_1, close_enough ) );
+        }
+    }
```

Comment:
This should also have (and support) rangeish types that don't have member begin (e.g. they provide ADL-findable begin only), and have different types for begin and end (the support for these is currently spotty, but the goal is to provide it eventually).

Honestly the test utils for this could use some refactoring and cleanup in general, so I don't blame you if you don't want to write the tests, but the impl should reasonably support these.